Check for invalid characters in name and value when setting a cookie

RupinMittal opened the pull request:
#### ea42e80465ced88ed7a4d396eb4aacffcae500f0
<pre>
Check for invalid characters in name and value when setting a cookie
<a href="https://bugs.webkit.org/show_bug.cgi?id=279151">https://bugs.webkit.org/show_bug.cgi?id=279151</a>
<a href="https://rdar.apple.com/135304982">rdar://135304982</a>

Reviewed by Chris Dumez.

As per the CookieStore API spec (<a href="https://wicg.github.io/cookie-store/#set-a-cookie)">https://wicg.github.io/cookie-store/#set-a-cookie)</a>,
we must return a type error when setting a cookie if:
(1) Certain invalid characters are in the name or value
(2) Name is empty and Value contains a &apos;=&apos; (signifies that user is trying to set
    both the name and value in the value field with &quot;name=value&quot; syntax)
(3) Name and Value are both empty

Currently, the WPT tests only test (2). This test was passing before this change--but
for the wrong reason. It was not passing because the CookieStore returns a type error
on empty name and &apos;=&apos; in value, but rather because the CookieStore allows this call and
eventually reaches the CFNetwork function to set a cookie--which incorrectly rejects
the request to set a cookie with an empty name (a bug has been filed for this).

This test for (2) now passes for the right reason, and we also add tests for (1) and (3).
A future patch will add them to the WPT repo as well.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any.js:
(promise_test.async testCase):
(promise_test.async testCase.async const):
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any.serviceworker-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::containsInvalidCharacters):
(WebCore::CookieStore::set):

Canonical link: <a href="https://commits.webkit.org/283624@main">https://commits.webkit.org/283624@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36a3c5425908d5d9b24f25104f54aede7732025b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66793 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46168 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19415 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70828 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17926 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68911 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53967 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17701 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53505 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12065 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69860 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42483 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57785 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34140 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39155 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15177 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16280 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61067 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15518 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72529 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10750 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60916 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10782 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57842 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61150 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8830 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2447 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10142 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41975 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43052 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44235 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42795 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -230,6 +230,18 @@ CookieStore::~CookieStore()
 
+static bool containsInvalidCharacters(const String& str)
+{
+    for (unsigned i = 0; i < str.length(); i++) {
```
I think this can use `const auto& character : str`

cdumez replied:
`auto&` without the `const` is more idiomatic in WebKit

RupinMittal replied:
I changed it to use `String::contains` as recommended below.

rr-codes commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -230,6 +230,18 @@ CookieStore::~CookieStore()
+{
+    for (unsigned i = 0; i < str.length(); i++) {
+        auto currentChar = str[i];
+        if (currentChar == 59 || currentChar == 127)
```
I think we tend to prefer to not have "magic" values like these, and instead have named constants.

RupinMittal replied:
That's a good point--I changed it to hex as recommended below--and since these are simply unicode character codes, I added a comment explaining where they come from.

rr-codes commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -392,6 +404,23 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
     cookie.name = WTFMove(options.name);
     cookie.value = WTFMove(options.value);
+
+    if (!cookie.name.length()) {
```
I think you can use `isEmpty` here and below

RupinMittal replied:
Fixed!

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any.js:
```diff
@@ -28,6 +28,11 @@ promise_test(async testCase => {
   assert_equals(cookie.value, 'cookie-value');
 }, 'cookieStore.set with name and value in options');
 
+promise_test(async testCase => {
```
please make sure you upstream these changes to WPT afterwards.

RupinMittal replied:
Will do!

cdumez replied:
Also, are the new tests passing in Chrome?

RupinMittal replied:
Yes!

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -230,6 +230,18 @@ CookieStore::~CookieStore()
     m_mainThreadBridge->detach();
 }
 
+static bool containsInvalidCharacters(const String& str)
```
I think you can call `String::contains()` which takes in a lambda

RupinMittal replied:
That does look cleaner--thank you!

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -230,6 +230,18 @@ CookieStore::~CookieStore()
+{
+    for (unsigned i = 0; i < str.length(); i++) {
+        auto currentChar = str[i];
+        if (currentChar == 59 || currentChar == 127)
```
We usually use hexadecimal to match the spec, e.g. `0x003B`, instead of decimal values.

RupinMittal replied:
Fixed with a comment to the spec so the numbers seem less random.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -230,6 +230,16 @@ CookieStore::~CookieStore()
+{
+    // The invalid characters are specified at https://wicg.github.io/cookie-store/#set-a-cookie.
+    return str.contains([](UChar c) {
+        return c == 0x003B
```
Could all these 3 conditions be on 1 line.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -392,6 +402,23 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+        }
+    }
+
+    if (containsInvalidCharacters(cookie.name) || containsInvalidCharacters(cookie.value)) {
```
It would be nice to split these 2 checks so we can provide a better promise rejection message.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -392,6 +402,23 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
     cookie.name = WTFMove(options.name);
     cookie.value = WTFMove(options.value);
+
+    if (cookie.name.isEmpty()) {
```
Isn't the spec not doing the `=` and special characters in the opposite order? This is web observable so ordering matters.

RupinMittal replied:
You're right--the ordering is 
1. Invalid characters
2. empty name and '=' in value
3. empty name and empty value

I've fixed this in the latest iteration.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -230,6 +230,14 @@ CookieStore::~CookieStore()
     m_mainThreadBridge->detach();
 }
 
+static bool containsInvalidCharacters(const String& str)
```
Please renaming `str` to `string`, no abbreviations in WebKit.

RupinMittal replied:
Fixed!

RupinMittal committed:
Check for invalid characters in name and value when setting a cookie
https://bugs.webkit.org/show_bug.cgi?id=279151
rdar://135304982

Reviewed by Chris Dumez.

As per the CookieStore API spec (https://wicg.github.io/cookie-store/#set-a-cookie),
we must return a type error when setting a cookie if:
(1) Certain invalid characters are in the name or value
(2) Name is empty and Value contains a '=' (signifies that user is trying to set
    both the name and value in the value field with "name=value" syntax)
(3) Name and Value are both empty

Currently, the WPT tests only test (2). This test was passing before this change--but
for the wrong reason. It was not passing because the CookieStore returns a type error
on empty name and '=' in value, but rather because the CookieStore allows this call and
eventually reaches the CFNetwork function to set a cookie--which incorrectly rejects
the request to set a cookie with an empty name (a bug has been filed for this).

This test for (2) now passes for the right reason, and we also add tests for (1) and (3).
A future patch will add them to the WPT repo as well.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any.js:
(promise_test.async testCase):
(promise_test.async testCase.async const):
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any.serviceworker-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::containsInvalidCharacters):
(WebCore::CookieStore::set):

Canonical link: https://commits.webkit.org/283624@main
